wally-registry-backend/src/auth.rs

```diff
@@ -91,10 +172,11 @@ impl<'r> FromRequest<'r> for WriteAccess {
                 Status::Unauthorized,
                 format_err!("Invalid API key for write access"),
             )),
-            AuthMode::ApiKey(key) => match_api_key(request, key, Self { _dummy: 0 }),
+            AuthMode::ApiKey(key) => match_api_key(request, key, Self { github: None }),
             AuthMode::DoubleApiKey { write, .. } => {
-                match_api_key(request, write, Self { _dummy: 0 })
+                match_api_key(request, write, Self { github: None })
             }
+            AuthMode::GithubOAuth => verify_github_token(request).await,
```

Comment:
Having `github: None` in all non-GitHub auth cases is a little funny. I think we've got a couple options to make this struct fit our intent a little better.

I think my favorite is to make `WriteAccess` an enum:
```rust
pub enum WriteAccess {
    /// Granted from a global API key; the user can do whatever they want!
    Global,

    /// Granted from a GitHub login; the user has limited scope
    GitHub(GitHubInfo)
}
```

This makes our matches a little more semantically clear, as a nice bonus.

Reply:
I forget how nice Rust enums are! I've used `Global` to indicate everyone has this access and `ApiKey` to indicate if they have access because of an api key. This makes sense to do as I've changed `ReadAccess` to also follow this. `WriteAccess` doesn't have a Global in this case as we intentionally don't support that. Doing this has really cleaned up stuff semantically and this is all feeling incredibly clean now!

Reply:
Actually I've decided to use `Public` instead of `Global`. I realised you probably meant global as in "access to all packages" so my usage could be confusing.